pkg/kube/informerfactory/factory.go

```diff
@@ -0,0 +1,222 @@
+// Additionally, aggregate operations like Start, Shutdown, and Wait are available.
+// Kubernetes core has informer factories with very similar logic. However, this has a few problems that
+// spurred a fork:
+// * Factories are per package. That means we have ~6 distinct factories, which makes management a hassle.
+// * Across these, the factories are often inconsistent in functionality. Changes to these takes >4 months.
+// * Lack of functionality we want (see below).
+//
+// Added functionality:
+// * Single factory for any type, including dynamic informers, meta informers, typed informers, etc.
+// * Ability to create multiple informers of the same type but with different filters.
+// * Ability to run a single informer rather than all of them.
+package informerfactory
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/pkg/kube/kubetypes"
+	"istio.io/istio/pkg/util/sets"
+)
+
+// NewInformerFunc returns a SharedIndexInformer.
+type NewInformerFunc func() cache.SharedIndexInformer
+
+type StartableInformer struct {
+	Informer cache.SharedIndexInformer
+	start    func(stopCh <-chan struct{})
+}
+
+func (s StartableInformer) Start(stopCh <-chan struct{}) {
+	s.start(stopCh)
+}
+
+// InformerFactory provides access to a shared informer factory
+type InformerFactory interface {
+	// Start initializes all requested informers. They are handled in goroutines
+	// which run until the stop channel gets closed.
+	Start(stopCh <-chan struct{})
+
+	// InformerFor returns the SharedIndexInformer the provided type.
+	InformerFor(resource schema.GroupVersionResource, opts kubetypes.InformerOptions, newFunc NewInformerFunc) StartableInformer
+
+	// WaitForCacheSync blocks until all started informers' caches were synced
+	// or the stop channel gets closed.
+	WaitForCacheSync(stopCh <-chan struct{}) bool
+
+	// Shutdown marks a factory as shutting down. At that point no new
+	// informers can be started anymore and Start will return without
+	// doing anything.
+	//
+	// In addition, Shutdown blocks until all goroutines have terminated. For that
+	// to happen, the close channel(s) that they were started with must be closed,
+	// either before Shutdown gets called or while it is waiting.
+	//
+	// Shutdown may be called multiple times, even concurrently. All such calls will
+	// block until all goroutines have terminated.
+	Shutdown()
+}
+
+// NewSharedInformerFactory constructs a new instance of informerFactory for all namespaces.
+func NewSharedInformerFactory() InformerFactory {
+	return &informerFactory{
+		informers:        map[informerKey]cache.SharedIndexInformer{},
+		startedInformers: sets.New[informerKey](),
+	}
+}
+
+// InformerKey represents a unique informer
+type informerKey struct {
+	gvr           schema.GroupVersionResource
+	labelSelector string
+	fieldSelector string
+	informerType  kubetypes.InformerType
```

Comment:
maybe type is not neccessary as gvr is there

Reply:
We actually end up creating 2 watches against the same resource (CRD - one full and one metadata) in some cases. This case is in the experimental off-by-default analysis code, so not a big issue.

If we remove this, you will actually get a panic on `istioctl analyze` since it will try to cast a CRD to a PartialMetadataObject or vis-versa